File: bindings/dart/lib/src/ldp.dart
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// LDP protocol types for proven-servers.

/// ContainerType matching the Idris2 ABI tags.
enum ContainerType {
  basic(0),
  direct(1),
  indirect(2);

  const ContainerType(this.tag);
  final int tag;

  static ContainerType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// LdpResourceType matching the Idris2 ABI tags.
enum LdpResourceType {
  rdfSource(0),
  nonRdfSource(1),
  container(2);

  const LdpResourceType(this.tag);
  final int tag;

  static LdpResourceType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// Preference matching the Idris2 ABI tags.
enum Preference {
  minimalContainer(0),
  includeContainment(1),
  includeMembership(2),
  omitContainment(3),
  omitMembership(4);

  const Preference(this.tag);
  final int tag;

  static Preference? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// InteractionModel matching the Idris2 ABI tags.
enum InteractionModel {
  ldpr(0),
  ldpc(1),
  ldpBasicContainer(2),
  ldpDirectContainer(3),
  ldpIndirectContainer(4);

  const InteractionModel(this.tag);
  final int tag;

  static InteractionModel? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ConstraintViolation matching the Idris2 ABI tags.
enum ConstraintViolation {
  membershipConstant(0),
  containsTriplesModified(1),
  serverManaged(2),
  typeConflict(3);

  const ConstraintViolation(this.tag);
  final int tag;

  static ConstraintViolation? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

File: bindings/javascript/src/ldp.js
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// LDP protocol types for proven-servers.

/** ContainerType matching the Idris2 ABI tags. */
export const ContainerType = Object.freeze({
  BASIC: 0,
  DIRECT: 1,
  INDIRECT: 2,
});

/** LdpResourceType matching the Idris2 ABI tags. */
export const LdpResourceType = Object.freeze({
  RDF_SOURCE: 0,
  NON_RDF_SOURCE: 1,
  CONTAINER: 2,
});

/** Preference matching the Idris2 ABI tags. */
export const Preference = Object.freeze({
  MINIMAL_CONTAINER: 0,
  INCLUDE_CONTAINMENT: 1,
  INCLUDE_MEMBERSHIP: 2,
  OMIT_CONTAINMENT: 3,
  OMIT_MEMBERSHIP: 4,
});

/** InteractionModel matching the Idris2 ABI tags. */
export const InteractionModel = Object.freeze({
  LDPR: 0,
  LDPC: 1,
  LDP_BASIC_CONTAINER: 2,
  LDP_DIRECT_CONTAINER: 3,
  LDP_INDIRECT_CONTAINER: 4,
});

/** ConstraintViolation matching the Idris2 ABI tags. */
export const ConstraintViolation = Object.freeze({
  MEMBERSHIP_CONSTANT: 0,
  CONTAINS_TRIPLES_MODIFIED: 1,
  SERVER_MANAGED: 2,
  TYPE_CONFLICT: 3,
});
